File: algorithms/code/go/sorting_algorithms.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// SortingStats holds statistics for sorting operations
type SortingStats struct {
	Comparisons int
	Swaps       int
	TimeElapsed time.Duration
}

// SortingAlgorithms provides various sorting algorithm implementations
type SortingAlgorithms struct{}

// NewSortingAlgorithms creates a new instance
func NewSortingAlgorithms() *SortingAlgorithms {
	return &SortingAlgorithms{}
}

// BubbleSort implements bubble sort algorithm
func (sa *SortingAlgorithms) BubbleSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	n := len(arr)
	arrCopy := make([]int, n)
	copy(arrCopy, arr)
	
	for i := 0; i < n-1; i++ {
		swapped := false
		for j := 0; j < n-i-1; j++ {
			stats.Comparisons++
			if arrCopy[j] > arrCopy[j+1] {
				arrCopy[j], arrCopy[j+1] = arrCopy[j+1], arrCopy[j]
				stats.Swaps++
				swapped = true
			}
		}
		if !swapped {
			break // Early termination if no swaps occurred
		}
	}
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// SelectionSort implements selection sort algorithm
func (sa *SortingAlgorithms) SelectionSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	n := len(arr)
	arrCopy := make([]int, n)
	copy(arrCopy, arr)
	
	for i := 0; i < n-1; i++ {
		minIdx := i
		for j := i + 1; j < n; j++ {
			stats.Comparisons++
			if arrCopy[j] < arrCopy[minIdx] {
				minIdx = j
			}
		}
		if minIdx != i {
			arrCopy[i], arrCopy[minIdx] = arrCopy[minIdx], arrCopy[i]
			stats.Swaps++
		}
	}
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// InsertionSort implements insertion sort algorithm
func (sa *SortingAlgorithms) InsertionSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	n := len(arr)
	arrCopy := make([]int, n)
	copy(arrCopy, arr)
	
	for i := 1; i < n; i++ {
		key := arrCopy[i]
		j := i - 1
		
		for j >= 0 {
			stats.Comparisons++
			if arrCopy[j] > key {
				arrCopy[j+1] = arrCopy[j]
				stats.Swaps++
				j--
			} else {
				break
			}
		}
		arrCopy[j+1] = key
	}
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// MergeSort implements merge sort algorithm
func (sa *SortingAlgorithms) MergeSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	arrCopy := make([]int, len(arr))
	copy(arrCopy, arr)
	
	sa.mergeSortHelper(arrCopy, 0, len(arrCopy)-1, &stats)
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// mergeSortHelper is a helper function for merge sort
func (sa *SortingAlgorithms) mergeSortHelper(arr []int, left, right int, stats *SortingStats) {
	if left < right {
		mid := left + (right-left)/2
		sa.mergeSortHelper(arr, left, mid, stats)
		sa.mergeSortHelper(arr, mid+1, right, stats)
		sa.merge(arr, left, mid, right, stats)
	}
}

// merge merges two sorted subarrays
func (sa *SortingAlgorithms) merge(arr []int, left, mid, right int, stats *SortingStats) {
	n1 := mid - left + 1
	n2 := right - mid
	
	leftArr := make([]int, n1)
	rightArr := make([]int, n2)
	
	for i := 0; i < n1; i++ {
		leftArr[i] = arr[left+i]
	}
	for j := 0; j < n2; j++ {
		rightArr[j] = arr[mid+1+j]
	}
	
	i, j, k := 0, 0, left
	
	for i < n1 && j < n2 {
		stats.Comparisons++
		if leftArr[i] <= rightArr[j] {
			arr[k] = leftArr[i]
			i++
		} else {
			arr[k] = rightArr[j]
			j++
		}
		k++
	}
	
	for i < n1 {
		arr[k] = leftArr[i]
		i++
		k++
	}
	
	for j < n2 {
		arr[k] = rightArr[j]
		j++
		k++
	}
}

// QuickSort implements quick sort algorithm
func (sa *SortingAlgorithms) QuickSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	arrCopy := make([]int, len(arr))
	copy(arrCopy, arr)
	
	sa.quickSortHelper(arrCopy, 0, len(arrCopy)-1, &stats)
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// quickSortHelper is a helper function for quick sort
func (sa *SortingAlgorithms) quickSortHelper(arr []int, low, high int, stats *SortingStats) {
	if low < high {
		pi := sa.partition(arr, low, high, stats)
		sa.quickSortHelper(arr, low, pi-1, stats)
		sa.quickSortHelper(arr, pi+1, high, stats)
	}
}

// partition partitions the array for quick sort
func (sa *SortingAlgorithms) partition(arr []int, low, high int, stats *SortingStats) int {
	pivot := arr[high]
	i := low - 1
	
	for j := low; j < high; j++ {
		stats.Comparisons++
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
			stats.Swaps++
		}
	}
	
	arr[i+1], arr[high] = arr[high], arr[i+1]
	stats.Swaps++
	return i + 1
}

// HeapSort implements heap sort algorithm
func (sa *SortingAlgorithms) HeapSort(arr []int) SortingStats {
	stats := SortingStats{}
	start := time.Now()
	
	arrCopy := make([]int, len(arr))
	copy(arrCopy, arr)
	
	n := len(arrCopy)
	
	// Build max heap
	for i := n/2 - 1; i >= 0; i-- {
		sa.heapify(arrCopy, n, i, &stats)
	}
	
	// Extract elements from heap one by one
	for i := n - 1; i > 0; i-- {
		arrCopy[0], arrCopy[i] = arrCopy[i], arrCopy[0]
		stats.Swaps++
		sa.heapify(arrCopy, i, 0, &stats)
	}
	
	stats.TimeElapsed = time.Since(start)
	return stats
}

// heapify maintains the heap property
func (sa *SortingAlgorithms) heapify(arr []int, n, i int, stats *SortingStats) {
	largest := i
	left := 2*i + 1
	right := 2*i + 2
	
	if left < n {
		stats.Comparisons++
		if arr[left] > arr[largest] {
			largest = left
		}
	}
	
	if right < n {
		stats.Comparisons++
		if arr[right] > arr[largest] {
			largest = right
		}
	}
	
	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		stats.Swaps++
		sa.heapify(arr, n, largest, stats)
	}
}

// GenerateRandomArray creates a random array of specified size
func (sa *SortingAlgorithms) GenerateRandomArray(size int) []int {
	arr := make([]int, size)
	for i := 0; i < size; i++ {
		arr[i] = rand.Intn(1000) + 1
	}
	return arr
}

// GenerateSortedArray creates a sorted array of specified size
func (sa *SortingAlgorithms) GenerateSortedArray(size int) []int {
	arr := make([]int, size)
	for i := 0; i < size; i++ {
		arr[i] = i + 1
	}
	return arr
}

// GenerateReverseSortedArray creates a reverse sorted array of specified size
func (sa *SortingAlgorithms) GenerateReverseSortedArray(size int) []int {
	arr := make([]int, size)
	for i := 0; i < size; i++ {
		arr[i] = size - i
	}
	return arr
}

// IsSorted checks if an array is sorted
func (sa *SortingAlgorithms) IsSorted(arr []int) bool {
	for i := 1; i < len(arr); i++ {
		if arr[i] < arr[i-1] {
			return false
		}
	}
	return true
}

// CompareAlgorithms compares all sorting algorithms on the same data
func (sa *SortingAlgorithms) CompareAlgorithms(arr []int) map[string]SortingStats {
	results := make(map[string]SortingStats)
	
	results["Bubble Sort"] = sa.BubbleSort(arr)
	results["Selection Sort"] = sa.SelectionSort(arr)
	results["Insertion Sort"] = sa.InsertionSort(arr)
	results["Merge Sort"] = sa.MergeSort(arr)
	results["Quick Sort"] = sa.QuickSort(arr)
	results["Heap Sort"] = sa.HeapSort(arr)
	
	return results
}

// PrintStats prints sorting statistics in a formatted way
func (sa *SortingAlgorithms) PrintStats(algorithm string, stats SortingStats) {
	fmt.Printf("%-15s: Comparisons=%d, Swaps=%d, Time=%v\n",
		algorithm, stats.Comparisons, stats.Swaps, stats.TimeElapsed)
}

// Example usage and demonstration
func main() {
	// Seed random number generator
	rand.Seed(time.Now().UnixNano())
	
	sa := NewSortingAlgorithms()
	
	// Test with different array sizes and types
	testCases := []struct {
		name string
		arr  []int
	}{
		{"Small Random", sa.GenerateRandomArray(10)},
		{"Medium Random", sa.GenerateRandomArray(100)},
		{"Large Random", sa.GenerateRandomArray(1000)},
		{"Sorted Array", sa.GenerateSortedArray(100)},
		{"Reverse Sorted", sa.GenerateReverseSortedArray(100)},
	}
	
	for _, tc := range testCases {
		fmt.Printf("\n=== %s (Size: %d) ===\n", tc.name, len(tc.arr))
		
		results := sa.CompareAlgorithms(tc.arr)
		
		for algorithm, stats := range results {
			sa.PrintStats(algorithm, stats)
		}
	}
	
	// Demonstrate individual algorithm
	fmt.Println("\n=== Individual Algorithm Demo ===")
	testArr := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Printf("Original array: %v\n", testArr)
	
	stats := sa.QuickSort(testArr)
	fmt.Printf("After Quick Sort: %v\n", testArr)
	sa.PrintStats("Quick Sort", stats)
	fmt.Printf("Is sorted: %t\n", sa.IsSorted(testArr))
}

File: algorithms/code/java/SortingAlgorithms.java
import java.util.*;

/**
 * Comprehensive implementation of various sorting algorithms in Java
 * Includes performance statistics and comparison capabilities
 */
public class SortingAlgorithms {
    
    /**
     * SortingStats holds statistics for sorting operations
     */
    public static class SortingStats {
        public int comparisons;
        public int swaps;
        public long timeElapsed; // in nanoseconds
        
        public SortingStats() {
            this.comparisons = 0;
            this.swaps = 0;
            this.timeElapsed = 0;
        }
        
        @Override
        public String toString() {
            return String.format("Comparisons=%d, Swaps=%d, Time=%d ns", 
                               comparisons, swaps, timeElapsed);
        }
    }
    
    /**
     * Bubble Sort implementation
     * Time Complexity: O(n²) worst case, O(n) best case
     * Space Complexity: O(1)
     */
    public SortingStats bubbleSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int n = arr.length;
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        
        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                stats.comparisons++;
                if (arrCopy[j] > arrCopy[j + 1]) {
                    swap(arrCopy, j, j + 1);
                    stats.swaps++;
                    swapped = true;
                }
            }
            if (!swapped) {
                break; // Early termination if no swaps occurred
            }
        }
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Selection Sort implementation
     * Time Complexity: O(n²) in all cases
     * Space Complexity: O(1)
     */
    public SortingStats selectionSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int n = arr.length;
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                stats.comparisons++;
                if (arrCopy[j] < arrCopy[minIdx]) {
                    minIdx = j;
                }
            }
            if (minIdx != i) {
                swap(arrCopy, i, minIdx);
                stats.swaps++;
            }
        }
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Insertion Sort implementation
     * Time Complexity: O(n²) worst case, O(n) best case
     * Space Complexity: O(1)
     */
    public SortingStats insertionSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int n = arr.length;
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        
        for (int i = 1; i < n; i++) {
            int key = arrCopy[i];
            int j = i - 1;
            
            while (j >= 0) {
                stats.comparisons++;
                if (arrCopy[j] > key) {
                    arrCopy[j + 1] = arrCopy[j];
                    stats.swaps++;
                    j--;
                } else {
                    break;
                }
            }
            arrCopy[j + 1] = key;
        }
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Merge Sort implementation
     * Time Complexity: O(n log n) in all cases
     * Space Complexity: O(n)
     */
    public SortingStats mergeSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        mergeSortHelper(arrCopy, 0, arrCopy.length - 1, stats);
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Helper method for merge sort
     */
    private void mergeSortHelper(int[] arr, int left, int right, SortingStats stats) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSortHelper(arr, left, mid, stats);
            mergeSortHelper(arr, mid + 1, right, stats);
            merge(arr, left, mid, right, stats);
        }
    }
    
    /**
     * Merge two sorted subarrays
     */
    private void merge(int[] arr, int left, int mid, int right, SortingStats stats) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        
        int[] leftArr = new int[n1];
        int[] rightArr = new int[n2];
        
        for (int i = 0; i < n1; i++) {
            leftArr[i] = arr[left + i];
        }
        for (int j = 0; j < n2; j++) {
            rightArr[j] = arr[mid + 1 + j];
        }
        
        int i = 0, j = 0, k = left;
        
        while (i < n1 && j < n2) {
            stats.comparisons++;
            if (leftArr[i] <= rightArr[j]) {
                arr[k] = leftArr[i];
                i++;
            } else {
                arr[k] = rightArr[j];
                j++;
            }
            k++;
        }
        
        while (i < n1) {
            arr[k] = leftArr[i];
            i++;
            k++;
        }
        
        while (j < n2) {
            arr[k] = rightArr[j];
            j++;
            k++;
        }
    }
    
    /**
     * Quick Sort implementation
     * Time Complexity: O(n log n) average case, O(n²) worst case
     * Space Complexity: O(log n) average case, O(n) worst case
     */
    public SortingStats quickSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        quickSortHelper(arrCopy, 0, arrCopy.length - 1, stats);
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Helper method for quick sort
     */
    private void quickSortHelper(int[] arr, int low, int high, SortingStats stats) {
        if (low < high) {
            int pi = partition(arr, low, high, stats);
            quickSortHelper(arr, low, pi - 1, stats);
            quickSortHelper(arr, pi + 1, high, stats);
        }
    }
    
    /**
     * Partition the array for quick sort
     */
    private int partition(int[] arr, int low, int high, SortingStats stats) {
        int pivot = arr[high];
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            stats.comparisons++;
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
                stats.swaps++;
            }
        }
        
        swap(arr, i + 1, high);
        stats.swaps++;
        return i + 1;
    }
    
    /**
     * Heap Sort implementation
     * Time Complexity: O(n log n) in all cases
     * Space Complexity: O(1)
     */
    public SortingStats heapSort(int[] arr) {
        SortingStats stats = new SortingStats();
        long startTime = System.nanoTime();
        
        int n = arr.length;
        int[] arrCopy = Arrays.copyOf(arr, arr.length);
        
        // Build max heap
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(arrCopy, n, i, stats);
        }
        
        // Extract elements from heap one by one
        for (int i = n - 1; i > 0; i--) {
            swap(arrCopy, 0, i);
            stats.swaps++;
            heapify(arrCopy, i, 0, stats);
        }
        
        stats.timeElapsed = System.nanoTime() - startTime;
        return stats;
    }
    
    /**
     * Maintain the heap property
     */
    private void heapify(int[] arr, int n, int i, SortingStats stats) {
        int largest = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        
        if (left < n) {
            stats.comparisons++;
            if (arr[left] > arr[largest]) {
                largest = left;
            }
        }
        
        if (right < n) {
            stats.comparisons++;
            if (arr[right] > arr[largest]) {
                largest = right;
            }
        }
        
        if (largest != i) {
            swap(arr, i, largest);
            stats.swaps++;
            heapify(arr, n, largest, stats);
        }
    }
    
    /**
     * Utility method to swap two elements in an array
     */
    private void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    
    /**
     * Generate a random array of specified size
     */
    public int[] generateRandomArray(int size) {
        Random random = new Random();
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = random.nextInt(1000) + 1;
        }
        return arr;
    }
    
    /**
     * Generate a sorted array of specified size
     */
    public int[] generateSortedArray(int size) {
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = i + 1;
        }
        return arr;
    }
    
    /**
     * Generate a reverse sorted array of specified size
     */
    public int[] generateReverseSortedArray(int size) {
        int[] arr = new int[size];
        for (int i = 0; i < size; i++) {
            arr[i] = size - i;
        }
        return arr;
    }
    
    /**
     * Check if an array is sorted
     */
    public boolean isSorted(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < arr[i - 1]) {
                return false;
            }
        }
        return true;
    }
    
    /**
     * Compare all sorting algorithms on the same data
     */
    public Map<String, SortingStats> compareAlgorithms(int[] arr) {
        Map<String, SortingStats> results = new HashMap<>();
        
        results.put("Bubble Sort", bubbleSort(arr));
        results.put("Selection Sort", selectionSort(arr));
        results.put("Insertion Sort", insertionSort(arr));
        results.put("Merge Sort", mergeSort(arr));
        results.put("Quick Sort", quickSort(arr));
        results.put("Heap Sort", heapSort(arr));
        
        return results;
    }
    
    /**
     * Print sorting statistics in a formatted way
     */
    public void printStats(String algorithm, SortingStats stats) {
        System.out.printf("%-15s: %s%n", algorithm, stats.toString());
    }
    
    /**
     * Main method for demonstration
     */
    public static void main(String[] args) {
        SortingAlgorithms sa = new SortingAlgorithms();
        
        // Test with different array sizes and types
        List<TestCase> testCases = Arrays.asList(
            new TestCase("Small Random", sa.generateRandomArray(10)),
            new TestCase("Medium Random", sa.generateRandomArray(100)),
            new TestCase("Large Random", sa.generateRandomArray(1000)),
            new TestCase("Sorted Array", sa.generateSortedArray(100)),
            new TestCase("Reverse Sorted", sa.generateReverseSortedArray(100))
        );
        
        for (TestCase tc : testCases) {
            System.out.printf("%n=== %s (Size: %d) ===%n", tc.name, tc.arr.length);
            
            Map<String, SortingStats> results = sa.compareAlgorithms(tc.arr);
            
            for (Map.Entry<String, SortingStats> entry : results.entrySet()) {
                sa.printStats(entry.getKey(), entry.getValue());
            }
        }
        
        // Demonstrate individual algorithm
        System.out.println("\n=== Individual Algorithm Demo ===");
        int[] testArr = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Original array: " + Arrays.toString(testArr));
        
        SortingStats stats = sa.quickSort(testArr);
        System.out.println("After Quick Sort: " + Arrays.toString(testArr));
        sa.printStats("Quick Sort", stats);
        System.out.println("Is sorted: " + sa.isSorted(testArr));
    }
    
    /**
     * Helper class for test cases
     */
    private static class TestCase {
        String name;
        int[] arr;
        
        TestCase(String name, int[] arr) {
            this.name = name;
            this.arr = arr;
        }
    }
}
